### ISSUE TITLE
Allow the processing of the ssh_CHANNEL_REQUEST only after response is sent

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8239]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8239|
|Type|enhancement|
|Created|2016-03-20 19:54:20Z|

As documented in RFC 4254 section 5.4 the client may request a reponse from SSH_MSG_CHANNEL_REQUEST.

While for the client-side behariou the RFC goes into details with

```

   The client is allowed to send further messages without waiting for
   the response to the request.
```

There are no explicit details whether the server is allowed to send other messages before the reply.

I would say that it make sense for the server to continue processing the request only after response is sent.

For example the Cisco SSH implementation is confused if after SSH_MSG_CHANNEL_REQUEST the session does not directly respond to the request and start sending other messages.

Under common usage this is not observed as for example SSHSession.request_exec will call a separate process which will need for another reactor loop so the response is sent before the actual command is executed.

But if the channel request is processed in memory without any other deferred the result of the action itself might be sent before the channel request confirmation.

```

cisco ssh   -> Send SSH_MSG_CHANNEL_REQUEST exec
twisted ssh -> Call SSHConnection.ssh_CHANNEL_REQUEST
                Call SSHChannel.requestReceived
                  Call SSHSession.request_exec(command_string)
                  Execute command_string in memory
                    Send MSG_CHANNEL_DATA
                    Send MSG_CHANNEL_EOF
                  Return SSHSession.request_exec with 1
                Return SSHChannel.requestReceived 1
                Send MSG_CHANNEL_SUCCESS 
```

```

cisco ssh   -> Send SSH_MSG_CHANNEL_REQUEST exec
twisted ssh -> Call SSHConnection.ssh_CHANNEL_REQUEST
                Call SSHChannel.requestReceived
                  Call SSHSession.request_exec(command_string)
                  Return SSHSession.request_exec with 1
                Return SSHChannel.requestReceived 1
                Send MSG_CHANNEL_SUCCESS 
                
                Call SSHChannel.continueRequset
                  Execute command_string in memory
                    Send MSG_CHANNEL_DATA
                    Send MSG_CHANNEL_EOF

```

I don't know if this is a bug or a feature.

What I know is that when integrating with Cisco SSH implementation extra care is required :)

<details><summary>Searchable metadata</summary>

```
trac-id__8239 8239
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__conch conch
keywords__None None
time__1458503660737074 1458503660737074
changetime__1458503660826903 1458503660826903
version__None None
owner__None None
cc__z3p
```
</details>
